File: leetcode/questions/346/python/346.py
from collections import deque
class MovingAverage:

    def __init__(self, size: int):
        self._deque = deque()
        self._size = size
        self._total_sum = 0

    def next(self, val: int) -> float:
        if len(self._deque) == self._size:
            to_remove = self._deque.popleft()
            self._total_sum -= to_remove

        self._deque.append(val)
        self._total_sum += val

        return self._total_sum / len(self._deque)


# Your MovingAverage object will be instantiated and called as such:
# obj = MovingAverage(size)
# param_1 = obj.next(val)

File: leetcode/questions/346/java/346.java
import java.util.ArrayDeque;

class MovingAverage {
    private ArrayDeque<Integer> deque;
    private int size;
    private double sum;

    public MovingAverage(int size) {
        this.deque = new ArrayDeque<>();
        this.size = size;
        this.sum = 0;
    }
    
    public double next(int val) {
        if (this.deque.size() == this.size) {
            this.sum -= this.deque.pollFirst();
        }
        this.deque.offerLast(val);
        this.sum += val;

        return this.sum / this.deque.size();
    }
}

/**
 * Your MovingAverage object will be instantiated and called as such:
 * MovingAverage obj = new MovingAverage(size);
 * double param_1 = obj.next(val);
 */